src/org/freenetproject/freemail/wot/WoTConnections.java

```diff
@@ -20,10 +20,399 @@
+        WoTConnectionImpl wotConnection;
+        if ((wotConnection = wotConnectionsPool.poll()) == null) {
+            wotConnection = new WoTConnectionImpl(pluginRespirator);
+        } else {
+            try {
+                wotConnection.ping();
+            } catch (IOException | WoTException ignored) {
+                if (wotConnectionsPool.size() > 0) {
+                    borrowWoTConnection();
+                } else {
+                    wotConnection = new WoTConnectionImpl(pluginRespirator);
+                    wotConnection.ping();
+                }
+            }
+        }
+
+        return wotConnection;
+    }
+
+    private void returnWoTConnection(WoTConnectionImpl wotConnection) {
+        if (wotConnection == null) {
+            return;
+        }
+
+        wotConnectionsPool.offer(wotConnection);
+    }
+
+    private static class WoTConnectionImpl implements WoTConnection {
+
+        private final FCPPluginConnection fcpPluginConnection;
+
+        private WoTConnectionImpl(PluginRespirator pr) throws PluginNotFoundException {
+            fcpPluginConnection = pr.connectToOtherPlugin(WoTProperties.WOT_FCP_NAME, new FredPluginFCPMessageHandler.ClientSideFCPMessageHandler() {
+                @Override
+                public FCPPluginMessage handlePluginFCPMessage(FCPPluginConnection connection, FCPPluginMessage message) {
+                    /* NOTICE: It is possible that the reply message <b>is</b> passed to the message handler
+                     * upon certain error conditions, for example if the timeout you specify when calling
+                     * sendSynchronous expires before the reply arrives. This is not guaranteed though.
+                     * - {@link FCPPluginConnection#sendSynchronous(FCPPluginMessage, long)} Javadoc */
+                    Logger.warning(this, "Receive unexpected FCPMessage from WoT connection: " + message);
+                    return null;
+                }
+            });
+        }
+
+        private void ping() throws IOException, WoTException, InterruptedException {
+            FCPPluginMessage message = FCPPluginMessage.construct();
+            message.params.putOverwrite("Message", "Ping");
+            FCPPluginMessage response = sendSynchronous(message, "Pong");
+            checkErrorMessage("ping", response);
+        }
+
+        @Override
+        public List<OwnIdentity> getAllOwnIdentities()
+                throws IOException, WoTException, InterruptedException {
+            FCPPluginMessage message = FCPPluginMessage.construct();
+            message.params.putOverwrite("Message", "GetOwnIdentities");
+            FCPPluginMessage response = sendSynchronous(message, "OwnIdentities");
+            checkErrorMessage("getAllOwnIdentities", response);
+
+            int identitiesAmount = Integer.parseInt(response.params.get("Amount"));
+            final List<OwnIdentity> ownIdentities = new ArrayList<>(identitiesAmount);
```

Comment:
`final` local variables are visual clutter, the JVM can figure this out on its own.